src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+"""
+Created on Sep 10, 2019
+@author: Ayelet Heimowitz, Amit Moscovich
+
+Integrated into ASPIRE by Garrett Wright Feb 2021.
+"""
+
+import logging
+import os
+
+import numpy as np
+from numpy import linalg as npla
+from scipy.optimize import linprog
+
+from aspire.image import Image
+from aspire.numeric import fft
+from aspire.utils import complex_type, eigs
+
+logger = logging.getLogger(__name__)
+
+
+def abs2(x):
+    """
+    Compute complex modulus squared.
+    """
+
+    return x.real ** 2 + x.imag ** 2
+
+
+class CtfEstimator:
+    """
+    CtfEstimator Class ...
+    """
+
+    def __init__(
+        self,
+        pixel_size,
+        cs,
+        amplitude_contrast,
+        voltage,
+        psd_size,
+        num_tapers,
+        dtype=np.float32,
+    ):
+        """
+        Instantiate a CtfEstimator instance.
+
+        :param pixel_size: Size of the pixel in Angstrom.
+        :param cs: Spherical aberration in mm.
+        :param amplitude_contrast: Amplitude contrast.
+        :param voltage: Voltage of electron microscope.
+        :param psd_size: Block size (in pixels) for PSD estimation.
+        :param num_tapers: Number of tapers to apply in PSD estimation.
+        :returns: CtfEstimator instance.
+        """
+
+        self.pixel_size = pixel_size
+        self.cs = cs
+        self.amplitude_contrast = amplitude_contrast
+        self.voltage = voltage
+        self.psd_size = psd_size
+        self.num_tapers = num_tapers
+        self.lmbd = 1.22639 / np.sqrt(voltage * 1000 + 0.97845 * np.square(voltage))
+        self.dtype = np.dtype(dtype)
+
+        center = psd_size // 2
+        [X, Y] = np.meshgrid(
+            np.arange(0 - center, psd_size - center, dtype=self.dtype) / psd_size,
+            np.arange(0 - center, psd_size - center, dtype=self.dtype) / psd_size,
+        )
+
+        rb = np.sqrt(np.square(X) + np.square(Y))
+
+        self.r_ctf = rb * (10 / pixel_size)
+        self.theta = np.arctan2(Y, X)
+        self.defocus1 = 0
+        self.defocus2 = 0
+        self.angle = 0
+        self.h = 0
+
+    def set_df1(self, df):
+        """
+        Sets defocus.
+
+        :param df: Defocus value in the direction perpendicular to df2.
+        """
+
+        self.defocus1 = df
+
+    def set_df2(self, df):
+        """
+        Sets defocus.
+
+        :param df: Defocus value in the direction perpendicular to df1.
+        """
+
+        self.defocus2 = df
+
+    def set_angle(self, angle):
+        """
+        Sets angle.
+
+        :param angle: Angle (in degrees) between df1 and the x-axis.
```

Comment:
Im not sure we should just convert on I/O because we have several functions that take in an angle (in degrees). I can convert everything, or leave it. Before I convert, just want to confirm that (converting whole to radians) is for the best.